vpr/src/route/four_ary_heap.cpp

```diff
@@ -0,0 +1,105 @@
+#include "four_ary_heap.h"
+#include "vtr_log.h"
+
+// The leftmost/smallest-index child of node i
+static inline size_t first_child(size_t i) { return (i << 2) - 2; }
+
+inline size_t FourAryHeap::parent(size_t i) const { return (i + 2) >> 2; }
+
+inline size_t FourAryHeap::smallest_child(size_t i) const {
+    // Returns first_child(i) if i has no children
+
+    size_t child_1 = first_child(i);
+    size_t child_2 = child_1 + 1;
+    size_t child_3 = child_1 + 2;
+    size_t child_4 = child_1 + 3;
+
+    size_t num_children = std::min((size_t)4, heap_tail_ - child_1);
+
+    switch (num_children) {
```

Comment:
I wonder if this is faster than a simple loop or slower than a loop?  We can merge this code, but I suggest running an experiment with this replaced by a loop and see if it is as fast or faster; it would be easier to read if there is no speed gain from this if nest.

Reply:
Using a loop is 3% slower; is this small enough to be worth changing?

Reply:
OK, let's leave as is then. If you can add a comment giving this result that would be good though.